### PR TITLE
Adds encryption_proxy example

### PR DESCRIPTION
**Pull request by heitorlessa**
Hi,

 This PR adds an encryption proxy example using current SAM version and it outputs both API Gateway entrypoints (RestApi) via Cloudformation.

 All details can be found in the README file - Please let me know if there are any questions. 

Thanks

### CODE REVIEW

File: examples/2016-10-31/encryption_proxy/README.MD [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+
+```bash
+aws cloudformation package --template template.yaml --s3-bucket <s3_bucket> --output-template packaged-template.yaml --region <aws_region>
+```
+
+Once complete, you should be ready to deploy it as a cloudformation stack:
+
+```bash
+aws cloudformation deploy --template packaged-template.yaml --stack-name encryption-proxy --capabilities CAPABILITY_IAM -- region <aws_region>
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/README.MD]: 1. --template-file
2. There is a space in --region

File: examples/2016-10-31/encryption_proxy/template.yaml [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            - 
+              Sid: "Allow Decryption Service to use this key"
+              Effect: "Allow"
+              Action:
+                - "kms:Decrypt"
+              Principal:
+                AWS: !GetAtt DecryptionServiceIAMRole.Arn
+              Resource: '*'
+
+    EncryptionServiceIAMRole:
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/template.yaml]: SAM Function resource will automatically create you a role with this permission through AWSLambdaBasicExecution managed policy. You can get rid of this.

File: examples/2016-10-31/encryption_proxy/template.yaml [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+                    -
+                      Sid: "AllowLambdaServiceToPushLogToCloudwatch"
+                      Effect: "Allow"
+                      Action:
+                        - "logs:CreateLogStream"
+                        - "logs:CreateLogGroup"
+                        - "logs:PutLogEvents"
+                      Resource: "arn:aws:logs:*:*:*"
+
+    DecryptionServiceIAMRole:
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/template.yaml]: You dont need this

File: examples/2016-10-31/encryption_proxy/src/decryption/index.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import boto3
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/src/decryption/index.py]: Can you move this to encryption_proxy/decryption.py? As much as we love directory structure, we'd like to keep things simple.

File: examples/2016-10-31/encryption_proxy/src/encryption/index.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import boto3
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/src/encryption/index.py]: Move to encryption_proxy/encryption.py please

File: examples/2016-10-31/encryption_proxy/template.yaml [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                        - "logs:CreateLogGroup"
+                        - "logs:PutLogEvents"
+                      Resource: "arn:aws:logs:*:*:*"
+
+    EncryptionFunction:
+      Type: 'AWS::Serverless::Function'
+      Properties:
+        Handler: index.post
+        Runtime: python2.7
+        CodeUri: ./src/encryption
```
Comment (sanathkr) [inline examples/2016-10-31/encryption_proxy/template.yaml]: To be consistent with other examples, set CodeUri to s3://<bucket-name>/encryption_proxy.zip

File: examples/2016-10-31/encryption_proxy/README.MD [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    - KMS Key ID is then passed to Encryption Lambda via Env vars
+
+
+#### How to launch it
+
+Firstly, you need to package both functions locally to S3 via the following command:
+
+
+```bash
+aws cloudformation package --template template.yaml --s3-bucket <s3_bucket> --output-template packaged-template.yaml --region <aws_region>
```
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/README.MD]: done

File: examples/2016-10-31/encryption_proxy/README.MD (hunk 1 above)
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/README.MD]: done

File: examples/2016-10-31/encryption_proxy/template.yaml [hunk 8]
```diff
@@ -0,0 +1,167 @@
+AWSTemplateFormatVersion: '2010-09-09'
+Transform: 'AWS::Serverless-2016-10-31'
+Description: >
+
+    This flourish example creates the following resources:
```
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/template.yaml]: done

File: examples/2016-10-31/encryption_proxy/template.yaml (hunk 3 above)
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/template.yaml]: Cleaned that up by making use of Managed Policy as follows:

```yaml
    EncryptionServiceIAMRole:
        Type: "AWS::IAM::Role"
        Properties:
            Path: "/"
            ManagedPolicyArns:
                - "arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
            ....
```

File: examples/2016-10-31/encryption_proxy/template.yaml (hunk 6 above)
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/template.yaml]: done -- Will send another commit to change README.MD to reflect that

File: examples/2016-10-31/encryption_proxy/template.yaml [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+        Environment:
+          Variables:
+            keyId: !Ref EncryptionDecryptionKey
+
+    DecryptionFunction:
+      Type: 'AWS::Serverless::Function'
+      Properties:
+        Handler: index.post
+        Runtime: python2.7
+        CodeUri: ./src/decryption
```
Comment (heitorlessa) [inline examples/2016-10-31/encryption_proxy/template.yaml]: done

**Comment by heitorlessa** [comment]
@sanathkr thanks for all comments - Just made changes on what was required and commented on those that can't as it breaks the example.

Let me know if you need anything else

**Comment by heitorlessa** [comment]
Currently working on simplifying the Exception Handling as there are some edge cases on TypeError too found in my tests today.

**Comment by heitorlessa** [comment]
@sanathkr Implemented your last comment on raising exceptions as it's now simplified and it also caught 2 bad inputs I wasn't handling before.

Let me know what else is required to push this forward.

**Comment by sanathkr** [comment]
Thanks for adding this example!

**Comment by prestomation** [comment]
Hello! This example is incomplete. There is no AWS::Serverless::Api defined in the template, although it is referenced.

**Comment by heitorlessa** [comment]
Hi Preston,

Would you mind expand on that?

API Gateway is implicitly defined under Events, and its implicit resource
_explicitly_ references in Outputs section.

If you run a describe Stack after deploying a SAM template you’d be able to
see why this works.

Happy to clarify any further thoughts

**Comment by prestomation** [comment]
Hi,
Sorry yes my apology! I have been using SAM for quite some time and was not aware that the API Gateway would be created implicitly.
